cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,327 @@
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    // Sanity check to confirm that the final distribution is broker set aware.
+    ensureBrokerSetAware(clusterModel, optimizationOptions);
+    if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+      _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+    }
+    finish();
+  }
+
+  private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check to confirm that the final distribution is brokerSet aware.
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+
+    for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic().entrySet()) {
+      if (!excludedTopics.contains(partitionsByTopic.getKey())) {
+        List<Partition> partitions = partitionsByTopic.getValue();
+        Set<Broker> allBrokersForTopic = partitions.stream()
+                                                   .map(partition -> partition.partitionBrokers())
+                                                   .flatMap(brokers -> brokers.stream())
+                                                   .collect(Collectors.toSet());
+        // Checks if a topic's brokers do not all live in a single brokerSet
+        if (_brokersByBrokerSet.values().stream().noneMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
+          throw new OptimizationFailureException(
+              String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).", name(), partitionsByTopic.getKey(), allBrokersForTopic));
+        }
+      }
+    }
+  }
+
+  /**
+   * brokerSet-awareness violations can be resolved with replica movements.
+   *
+   * @param broker              Broker to be balanced.
+   * @param clusterModel        The state of the cluster.
+   * @param optimizedGoals      Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+
+    String currentBrokerSetId = _brokerSetResolutionHelper.brokerSetId(broker);
+
+    for (Replica replica : broker.trackedSortedReplicas(replicaSortName(this, false, false)).sortedReplicas(true)) {
+      String eligibleBrokerSetIdForReplica =
+          _replicaToBrokerSetMappingPolicy.brokerSetIdForReplica(replica, clusterModel, _brokerSetResolutionHelper);
```

Comment:
Open discussion:

Would it be better to store the topicToBrokerSet map at initGoalState?

Think about some cases that users have complicated policy. We don't want to calculate it on each rebalance. Instead we may want to get the fixed mapping at the goal initialization.

Reply:
Valid thought. I have been trying to keep it the same way as **CapacityResolver** does - it has **capacityForBroker** method. Precomputation also has a disadvantage where we don't fail fast and do computations for dead brokers. Since a single rebalance can fail with OptmizationFailureException, in some cases we will end up doing computations for the replicas which we would not need. The implementation can be smart itself and can take the responsibility of caching the values.

Reply:
Worth involving @efeg for this discussion. 

@mohitpali did you mean the policy itself can have some caches?

Reply:
yes, exactly